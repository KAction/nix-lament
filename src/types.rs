use anyhow;
use tree_sitter as ts;

/// Kind of the lamentations. This is necessary to run only requested subset of supported
/// lamentations.
#[derive(Debug)]
pub enum Kind {
    D001, // Both "pname" and "name" in a call to "mkDerivation"
}

// Struct that describes location of the problematic code.
pub struct Lamentation {
    // By convention, lines are counted from one, columns -- from zero.
    pub line: usize,
    pub column: usize,
    pub kind: Kind,
    pub message: String,
}

// "lament" is a method, not plain function so we have space for
// once-per-module initialization (e.g parsing queries).
pub trait Lament {
    fn lament(&self, tree: &ts::Tree, content: &[u8]) -> Vec<Lamentation>;
}

#[derive(Clone, Copy)]
pub struct Module {
    pub kinds: &'static [Kind],
    pub new: fn() -> anyhow::Result<Box<dyn Lament>>,
}
