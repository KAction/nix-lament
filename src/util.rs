use crate::types;
use anyhow;
use tree_sitter as ts;
use tree_sitter_nix as nix;

pub fn node_text<'a>(node: &ts::Node, content: &'a [u8]) -> &'a str {
    std::str::from_utf8(&content[node.start_byte()..node.end_byte()]).unwrap()
}

type Handler = fn(m: &ts::QueryMatch, content: &[u8]) -> Option<types::Lamentation>;
pub struct PerMatch {
    query: ts::Query,
    handler: Handler,
}

impl PerMatch {
    pub fn new(query: &'static str, handler: Handler) -> anyhow::Result<Box<dyn types::Lament>> {
        Ok(Box::new(Self {
            query: ts::Query::new(nix::language(), query)?,
            handler,
        }))
    }
}

impl types::Lament for PerMatch {
    fn lament(&self, tree: &ts::Tree, content: &[u8]) -> Vec<types::Lamentation> {
        let mut out = vec![];
        let mut cursor = ts::QueryCursor::new();

        for m in cursor.matches(&self.query, tree.root_node(), content) {
            match (self.handler)(&m, content) {
                Some(x) => out.push(x),
                None => (),
            }
        }
        out
    }
}

#[macro_export]
macro_rules! via_match (
    ($name:ident, $handler:expr) => {
        fn new() -> anyhow::Result<Box<dyn crate::types::Lament>> {
            crate::util::PerMatch::new(include_str!(concat!(stringify!($name), ".scm")), $handler)
        }
        pub static MODULE: crate::types::Module = crate::types::Module {
            new, kinds: &[crate::types::Kind::$name],
        };
    };
);
