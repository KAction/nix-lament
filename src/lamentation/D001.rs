use crate::types;
use crate::util::node_text;
use crate::via_match;

via_match!(D001, |m, content| {
    let func = node_text(&m.captures[0].node, content);
    let n2 = m.captures[2]; // second name|pname match
    let point = n2.node.start_position();
    let message = format!("Call to `{}' has both `name' and `pname' arguments.", func);

    Some(types::Lamentation {
        kind: types::Kind::D001,
        line: point.row,
        column: point.column,
        message,
    })
});
