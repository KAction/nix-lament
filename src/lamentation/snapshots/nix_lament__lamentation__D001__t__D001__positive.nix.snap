---
source: src/lamentation/D001.rs
expression: "w.lament(&tree, &content[..])"
---
- line: 4
  column: 2
  kind: D001
  message: "Call to `mkDerivation' has both `name' and `pname' arguments."

